config/nat.py

```diff
@@ -0,0 +1,1104 @@
+#
+@nat.group('remove')
+def remove():
+    """Remove NAT-related configutation tasks"""
+    pass
+
+#
+# 'nat set' group ('config nat set ...')
+#
+@nat.group('set')
+def set():
+    """Set NAT-related timeout configutation tasks"""
+    pass
+
+#
+# 'nat reset' group ('config nat reset ...')
+#
+@nat.group('reset')
+def reset():
+    """Reset NAT-related timeout configutation tasks"""
+    pass
+
+#
+# 'nat add static' group ('config nat add static ...')
+#
+@add.group('static')
+def static():
+    """Add Static related configutation"""
+    pass
+
+#
+# 'nat add static basic' command ('config nat add static basic <global-ip> <local-ip>')
+#
+@static.command('basic')
+@click.pass_context
+@click.argument('global_ip', metavar='<global_ip>', required=True)
+@click.argument('local_ip', metavar='<local_ip>', required=True)
+@click.option('-nat_type', metavar='<nat_type>', required=False, type=click.Choice(["snat", "dnat"]), help="Set nat type")
+@click.option('-twice_nat_id', metavar='<twice_nat_id>', required=False, type=click.IntRange(1, 9999), help="Set the twice nat id")
+def add_basic(ctx, global_ip, local_ip, nat_type, twice_nat_id):
+    """Add Static NAT-related configutation"""
+
+    # Verify the ip address format 
+    if is_valid_ipv4_address(local_ip) is False:
+        ctx.fail("Given local ip address {} is invalid. Please enter a valid local ip address !!".format(local_ip)) 
+
+    if is_valid_ipv4_address(global_ip) is False:
+        ctx.fail("Given global ip address {} is invalid. Please enter a valid global ip address !!".format(global_ip))
+   
+    config_db = ConfigDBConnector()
+    config_db.connect()
+
+    entryFound = False
+    table = "STATIC_NAT"
+    key = global_ip
+    dataKey1 = 'local_ip'
+    dataKey2 = 'nat_type'
+    dataKey3 = 'twice_nat_id'
+
+    data = config_db.get_entry(table, key)
+    if data:
+        if data[dataKey1] == local_ip:
+            click.echo("Trying to add static nat entry, which is already present.")
+            entryFound = True
+
+    if nat_type == 'snat':
+        ipAddress = local_ip
+    else:
+        ipAddress = global_ip
+
+    if isIpOverlappingWithAnyStaticEntry(ipAddress, 'STATIC_NAPT') is True:
+        ctx.fail("Given entry is overlapping with existing NAPT entry !!")
+
+    if isOverlappingWithAnyDynamicEntry(ipAddress) is True:
+        ctx.fail("Given entry is overlapping with existing Dynamic entry !!")
+
+    if entryFound is False:
+        counters_db = SonicV2Connector(host="127.0.0.1")
+        counters_db.connect(counters_db.COUNTERS_DB)
+        snat_entries = 0
+        exists = counters_db.exists(counters_db.COUNTERS_DB, 'COUNTERS_GLOBAL_NAT:Values')
+        if exists:
+            counter_entry = counters_db.get_all(counters_db.COUNTERS_DB, 'COUNTERS_GLOBAL_NAT:Values')
+            if 'SNAT_ENTRIES' in counter_entry:
+                snat_entries = counter_entry['SNAT_ENTRIES']
+
+        if int(snat_entries) >= 1024:
```

Comment:
This limit of 1k entries is for SNAT or DNAT or overall static nat entry limit ?
Is this a Software limit or hardware limit ?

Reply:
1K is hardware limitation only, Both SNAT and DNAT limits the entries to 1k.
Single Static NAT entry creates a SNAT and DNAT entry is hardware.

Reply:
Yes, Added "MAX_ENTRIES" in counter_db which will be queried from HW at natorch code.